repo/format/format_manager.go

```diff
@@ -0,0 +1,466 @@
+	"github.com/kopia/kopia/repo/hashing"
+	"github.com/kopia/kopia/repo/logging"
+)
+
+var log = logging.Module("kopia/repo/format")
+
+// UniqueIDLength is the length of random unique ID of each repository.
+const UniqueIDLength = 32
+
+// Manager manages the contents of `kopia.repository` and `kopia.blobcfg`.
+type Manager struct {
+	//nolint:containedctx
+	ctx           context.Context // +checklocksignore
+	blobs         blob.Storage    // +checklocksignore
+	validDuration time.Duration   // +checklocksignore
+	password      string          // +checklocksignore
+	cache         BlobCache       // +checklocksignore
+
+	timeNow func() time.Time // +checklocksignore
+
+	// all the stuff protected by a mutex is valid until `validUntil`
+	mu sync.RWMutex
+	// +checklocks:mu
+	formatEncryptionKey []byte
+	// +checklocks:mu
+	j *KopiaRepositoryJSON
+	// +checklocks:mu
+	repoConfig *RepositoryConfig
+	// +checklocks:mu
+	blobCfgBlob BlobStorageConfiguration
+	// +checklocks:mu
+	current Provider
+	// +checklocks:mu
+	validUntil time.Time
+	// +checklocks:mu
+	loadedTime time.Time
+	// +checklocks:mu
+	refreshCounter int
+}
+
+func (m *Manager) getFormat() Provider {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.current
+}
+
+func (m *Manager) getOrRefreshFormat() (Provider, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if err := m.maybeRefreshLocked(); err != nil {
+		return nil, err
+	}
+
+	return m.current, nil
+}
+
+// +checklocks:m.mu
+func (m *Manager) maybeRefreshLocked() error {
+	if n := m.timeNow(); !n.Before(m.validUntil) {
+		// current format not valid anymore, kick off a refresh
+		if err := m.refreshLocked(m.ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// readAndCacheRepositoryBlobBytes reads the provided blob from the repository or cache directory.
+func (m *Manager) readAndCacheRepositoryBlobBytes(ctx context.Context, blobID blob.ID) ([]byte, time.Time, error) {
+	if data, mtime, ok := m.cache.Get(ctx, blobID); ok {
+		// read from cache and still valid
+		age := m.timeNow().Sub(mtime)
+
+		if age < m.validDuration {
+			return data, mtime, nil
+		}
+	}
+
+	var b gather.WriteBuffer
+	defer b.Close()
+
+	if err := m.blobs.GetBlob(ctx, blobID, 0, -1, &b); err != nil {
+		return nil, time.Time{}, errors.Wrapf(err, "error getting %s blob", blobID)
+	}
+
+	data := b.ToByteSlice()
+
+	mtime, err := m.cache.Put(ctx, blobID, data)
+
+	return data, mtime, errors.Wrapf(err, "error adding %s blob", blobID)
+}
+
+// RefreshCount returns the number of time the format has been refreshed.
+func (m *Manager) RefreshCount() int {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.refreshCounter
+}
+
+// refreshLocked reads `kopia.repository` blob, potentially from cache and decodes it.
+// +checklocks:m.mu
+func (m *Manager) refreshLocked(ctx context.Context) error {
+	log(ctx).Infow("refreshLocked", "now", m.timeNow())
+
+	b, cacheMTime, err := m.readAndCacheRepositoryBlobBytes(ctx, KopiaRepositoryBlobID)
```

Comment:
Can remove the cache and instead keep (cache) the parsed forms (result of `ParseKopiaRepositoryJSON`, `addFormatBlobChecksumAndLength` and `blobCfg` in `Manager`).  `refreshLocked` compares current time with what was cache reload duration.
A session approach might be better for use of the `blobCfg` and the `repoConfig` - maintain values for an instance of a callpath instead of wallclock-time.

Reply:
the on-disk cache allows us to not fetch the blob over and over again when kopia is invoked from the CLI.

Happy to chat more about rationales here, ping me on slack.